src/sentry/api/endpoints/source_map_debug_blue_thunder_edition.py

```diff
@@ -238,34 +239,39 @@ def get(self, request: Request, project: Project, event_id: str) -> Response:
 
 
 def get_source_file_data(abs_path, project, release, event):
-    filenme_choices = ReleaseFile.normalize(abs_path)
+    matching_source_file_names = ReleaseFile.normalize(abs_path)
+    matching_source_map_name = None
+    source_map_reference = None
+    source_file_lookup_result = "unsuccessful"
+    source_map_lookup_result = "unsuccessful"
```

Comment:
m: can we codify lookup states + found source file into a state machine? That way there are only a finite set of possibilities for all of these values - easier to understand what is mutated where.

Can we also type return value? We can use dataclass.

its generally really hard to follow the nested loops/ifs with breaks.

Reply:
Do you mind explaining what you mean by state machine? I mean from a python point of view. How would you implement that?

Reply:
What's hard for me to understand as a reader is all the different permutations of the state variables here. Does `matching_source_map_name` always need to be defined? Or only if `source_map_lookup_result` is `found`?

It also helps that if something is considered `found`, it becomes an invariant to move it from `found` -> something else. that is not clearly defined in the current implementation.

A state machine was just the abstraction I picked that made sense to me, but thinking about it more I can also see a class that has clear helpers and prevents impossible states. You can just call `return source_file_data.to_dict()` or similar to return the expected data from this function.

```py
# has all initial params
matching_source_file_names = ReleaseFile.normalize(abs_path)
source_file_data = SourceFileData(matching_source_file_names)

// ...


```

Reply:
Ok I agree the logic was very convoluted in that function. I extracted the whole mechanism into a class now. It's still quite complicated but a bit more readable imo. I think the whole thing is complicated in nature.

Is this more or less what you were looking for?